File: Aug/Day18/24-game.cc
#include <vector>
#include <cmath>
using namespace std;

class Solution {
public:
    bool judgePoint24(vector<int>& cards) {
        vector<double> nums(cards.begin(), cards.end());
        return solve(nums);
    }

private:
    bool solve(vector<double> nums) {
        if (nums.size() == 1) {
            return fabs(nums[0] - 24) < 1e-6;
        }
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < nums.size(); j++) {
                if (i != j) {
                    vector<double> remaining;
                    for (int k = 0; k < nums.size(); k++) {
                        if (k != i && k != j) remaining.push_back(nums[k]);
                    }
                    vector<double> candidates = {
                        nums[i] + nums[j],
                        nums[i] - nums[j],
                        nums[j] - nums[i],
                        nums[i] * nums[j]
                    };
                    for (double val : candidates) {
                        vector<double> next = remaining;
                        next.push_back(val);
                        if (solve(next)) return true;
                    }
                    if (fabs(nums[j]) > 1e-6) {
                        vector<double> next = remaining;
                        next.push_back(nums[i] / nums[j]);
                        if (solve(next)) return true;
                    }
                    if (fabs(nums[i]) > 1e-6) {
                        vector<double> next = remaining;
                        next.push_back(nums[j] / nums[i]);
                        if (solve(next)) return true;
                    }
                }
            }
        }
        return false;
    }
};

File: Aug/Day18/24-game.py
class Solution(object):
    def judgePoint24(self, cards):
        def solve(nums):
            if len(nums) == 1:
                return abs(nums[0] - 24) < 1e-6
            for i in range(len(nums)):
                for j in range(len(nums)):
                    if i != j:
                        remainingNums = [nums[k] for k in range(len(nums)) if k != i and k != j]
                        for op in [nums[i]+nums[j], nums[i]-nums[j], nums[j]-nums[i], nums[i]*nums[j]]:
                            if solve(remainingNums + [op]):
                                return True
                        if abs(nums[j]) > 1e-6:
                            if solve(remainingNums + [nums[i]/nums[j]]):
                                return True
                        if abs(nums[i]) > 1e-6:
                            if solve(remainingNums + [nums[j]/nums[i]]):
                                return True
            return False
        return solve(list(map(float, cards)))